### PR TITLE
feat: add the sameplacementaffinity plugin (filtering)

### PR DESCRIPTION
**Pull request by zhiying-lin**
### Description of your changes

- add sameplacementantiaffinity plugin and implement the filter extension point

Fixes #

I have:

- [X] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

Added unit tests

### Special notes for your reviewer

Scoring will be sent out in a separate PR.

### CODE REVIEW

**Comment by michaelawyu** [comment]
Hi Zhiying!

Can we separate the minor fixes into a separate PR?

**Comment by michaelawyu** [comment]
Also, the sameplacementaffinity plugin needs to connect to the scoring stages.

**Comment by michaelawyu** [comment]
This requires a change of CycleState that I haven't merged into the code base yet; I'll send the PR later today.

**Comment by zhiying-lin** [comment]
> Can we separate the minor fixes into a separate PR?
sure, created https://github.com/Azure/fleet/pull/451

> Also, the sameplacementaffinity plugin needs to connect to the scoring stages.

I did not get this part. Why does the plugin need the scoring stage? is it because the scoring is required stage and has to be implemented?

**Comment by michaelawyu** [comment]
Hi Zhiying! I replied that in chat, but the reason is that earlier we introduced a BoundOrScheduled score which helps the scheduler to prioritize clusters that already has an associated binding

**Comment by michaelawyu** [comment]
Come to think of that it might be better to call it `PrevBoundOrScheduled` score :D

File: pkg/scheduler/framework/plugins/sameplacementaffinity/filtering_test.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+)
+
+func TestFilter(t *testing.T) {
+	tests := []struct {
+		name                     string
+		scheduledOrBoundBindings []*fleetv1beta1.ClusterResourceBinding
+		want                     *framework.Status
+	}{
+		{
+			name: "placement has already been scheduled",
```
Comment (michaelawyu) [inline pkg/scheduler/framework/plugins/sameplacementaffinity/filtering_test.go:39]: Just a nit, the name should be `placement has been unscheduled` I assume?

Comment (michaelawyu) [reply pkg/scheduler/framework/plugins/sameplacementaffinity/filtering_test.go:39]: (The matching binding is unscheduled)

**Review by michaelawyu — DISMISSED** [review]
Just a question regarding tests, others LGTM.

File: pkg/scheduler/framework/plugins/sameplacementaffinity/filtering_test.go (hunk 1 above)
Comment (zhiying-lin) [reply pkg/scheduler/framework/plugins/sameplacementaffinity/filtering_test.go:39]: The binding should be scheduled. Fixed.

**Pull request merged**